packages/features.lua

```diff
@@ -104,53 +111,146 @@ local value = C(SILE.parserBits.integer)
-      if not res then SU.error("Bad OpenType value "..v.." for feature "..k) end
-      if type(res) == "string" then
-        t[res] = {posneg = posneg}
+      local feats = {}
+      local m = lpeg.match(fontspeclist, v)
+
+      if m then
+        for i, match in pairs(m) do
+          feats[i] = match
+        end
       else
-        t[res.key] = { posneg = posneg, value = res.value}
+          feats[k] = v
+      end
+
+      SU.debug("features", string.format("Built feats: %s", SU.dumpString(false, feats)))
+
+      for _, vv in pairs(feats) do
+        local pk, pv = _parsefontfeaturevalue(k, vv)
+        t[pk] = pv
       end
     end
   end
+  SU.debug("features", string.format("Interpreted features %s as %s.",
+                                     SU.dumpString(false, options),
+                                     SU.dumpString(false, t)        ))
+  return t
+end
 
-  SILE.settings.set("font.features", table2featurestring(t))
+-- We do it this way so that we can use a SILE.temporarily in \font,
+-- instead of calling these user-facing functions in it.
+SILE.registerCommand("add-font-feature", function(options, _)
+    local t = parsefontfeatures(options, nil)
+    local features_s = table2featurestring(t)
+    SILE.settings.set("font.features", features_s)
+    SU.debug("features", string.format("Added feature(s) %s", features_s))
 end)
 
-SILE.registerCommand("remove-font-feature", function (options, _)
-  local t = featurestring2table(SILE.settings.get("font.features"))
+local removefontfeatures = function (options, _)
+    local t_cur = featurestring2table(SILE.settings.get("font.features"))
+    local t_rm = parsefontfeatures(options, nil)
+    local t = pl.tablex.deepcopy(t_cur)
 
-  for k, v in pairs(options) do
-    if not opentype[k] then SU.warn("Unknown Opentype feature "..k)
-    else
-      v = v:gsub("^No", "")
-      local res
-      if type(opentype[k]) == "function" then res = opentype[k](v) else res = opentype[k][v] end
-      if not res then SU.error("Bad OpenType value "..v.." for feature "..k) end
-      if type(res) == "string" then t[res] = nil else t[res.key] = nil end
+    for k, v in pairs(t_rm) do
+        -- \remove-font-features{Ligatures=NoHistoric} should not remove
+        -- Historic
+        if t[k] and t[k].posneg == v.posneg then t[k] = nil end
     end
-  end
 
-  SILE.settings.set("font.features", table2featurestring(t))
-end)
+    SILE.settings.set("font.features", table2featurestring(t))
+    SU.debug("features", string.format("Features were: %s\nRemoved features: %s\nFeatures are now: %s",
+                                        SU.dumpString(false, t_cur), SU.dumpString(false, t_rm),
+                                        SU.dumpString(false, t)))
+end
+SILE.registerCommand("remove-font-feature", removefontfeatures)
+
+local fontfn = SILE.Commands.font
+SILE.registerCommand("font", function(options, content)
```

Comment:
I don't think this should be replacing `\font` across all of SILE when loaded. There are some cases where changing a core function are okay (and certainly in your own projects you can do whatever) but I don't think replacing this is the right way to do it. If nothing else any changes we make to `\font` would need porting to more than one place. Instead use a different command and call out to the original where needed. Also easier to document this way!

Reply:
Did I [misunderstand @simoncozens?](https://github.com/sile-typesetter/sile/issues/992#issuecomment-665575353)

Reply:
I think so. He said "wraps" not "replaces".

Reply:
Oh, OK. Sorry, I'm dumb. :man_facepalming: I understood "wrap" to mean "function that wraps `\font`", not "command that wraps `\font`.

What should I call it? `\font-and-features`? `\font-features`? `\Font`?

Reply:
No, @ctrlcctrlv didn't misunderstand, and this is the approach I would have taken - save the old `\font` command and extend it. I'm happy for a package to *extend* a built-in, and using the approach of saving and then calling the old command wouldn't mean we need to port changes to `\font` here too. I seem to remember someone wrote some code to do this kind of command wrapping, which is what I had in mind, but I  can't find it now.

Reply:
My concern was less with whether in re-used the function name (although I'd suggest it would be a lot less confusing if it introduced a new function than replacing the original) and more with the fact that the entire body of `\font` was being re-implemented. Whatever you do with the command name I do think this needs to wrap the old font command.

`SILE.setCommandDefaults()` does something like this but in only designed to mess with default values for *options*, it won't work for this case. The technique is there though, just stash the old command, register a new one, parse the feature names down to a format the old `\font` will understand, then pass them along to the old command.

Reply:
> The technique needed is not that different though, just stash the old command, register a new one, use the new one to parse the feature names down to a format the old \font will understand, then pass them along to the old command.

Isn't that what this change does?

```
local fontfn = SILE.Commands.font
 SILE.registerCommand("font", function(options, content)
    -- Do stuff with options
    fontfn(options, content)
end)

Reply:
I apologize, I managed to mess up reading that bit in the code. Sorry to keep things so confusing, yes that is a wrap!

My remaining concern is that it's not possible to load the *features* package to use the other functions (e.g. `\add-feature`) without getting the re-mastered `\font` as part of the deal. I'm not sure how important that is or isn't, just something to think about.